File: Lasso - Ridge Regression/Computer_Data_Lasso_Ridge.py
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

comp = pd.read_csv('C:/Users/lnram/Desktop/ExcelR/Module 7 - Multiple Linear Regresion/Assignments/Computer_Data.csv')
comp.head()
comp = comp.iloc[:,1:] # Removing index column
# Business Problem - Predicting Price of Computer


# Checking for missing values
comp.isna().sum()  # No Missing values

# Summary Statistics
comp.describe()

# Different data Types
comp.dtypes # three categorical variables - cd, multi, premium


# Continous variables

# Skeweness
comp.skew()

#Kurtosis
comp.kurt()

# Graphical Analysis

comp.select_dtypes(exclude = "object").columns  # to get numerical columns

# Distribution Plots
sns.distplot(comp.price , hist = False , label = "price");
sns.distplot(comp.speed , hist = False , label = "speed"); # Multi modal
sns.distplot(comp.hd , hist = False , label = "hd"); # Multi modal and right skewed
sns.distplot(comp.ram , hist = False , label = "ram");  # Multi modal and right skewed
sns.distplot(comp.screen , hist = False , label = "screen"); # multi modal and right skewed
sns.distplot(comp.ads , hist = False , label = "ads"); # left skewed and bi modal
sns.distplot(comp.trend , hist = False , label = "trend"); # peakedness

comp.boxplot() # Outliers exists for Price, hd, ram, screen

# Discrete Analysis
comp.select_dtypes("object").columns

comp['cd'].value_counts(normalize=True) # Balanced data
comp['multi'].value_counts(normalize=True) # Imbalanced data
comp['premium'].value_counts(normalize=True) #Imbalanced data

# Relationship with Price
sns.pairplot(comp)
sns.heatmap(comp.corr() , annot = True) # Price and ram are moderately correlated, ram and hd are moderately correlated

# Model Building - Multiple Linear Regression
# Steps - Build model, Score, Predicting, RMSE, Model Assumptions

from sklearn.linear_model import LinearRegression
LR1 = LinearRegression()

# Converting categorical variables to dummies and dropping the first column to N-1 dummies
new_comp = pd.get_dummies(comp , columns=comp.select_dtypes("object").columns , drop_first = True)

x = new_comp.iloc[:,1:]
y = new_comp.iloc[:,0]


LR1.fit(x , y)
# Getting coefficients of variables               
LR1.coef_
LR1.intercept_

# Getting Adjusted R squared value
LR1.score(x,y)

pred1 = LR1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LR1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running Ridge Regression
from sklearn.linear_model import Ridge
RM1 = Ridge(alpha = 0.4,normalize=True)
RM1.fit(x,y)
# Coefficient values for all the independent variables
RM1.coef_
RM1.intercept_
plt.bar(height = pd.Series(RM1.coef_),x=pd.Series(x.columns))

# Getting Adjusted R squared value
RM1.score(x,y)

pred1 = RM1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LR1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running a Ridge Regressor of set of alpha values and observing how the R-Squared
rmse = []
R_sqrd = []
alphas = np.arange(0,100,0.05)
for i in alphas:
    RM = Ridge(alpha = i,normalize=True)
    RM.fit(x , y)
    R_sqrd.append(RM.score(x , y))
    rmse.append(np.sqrt(np.mean((RM.predict(x) - y)**2)))
    
# Plotting RMSE,R_Squared values with respect to alpha values

# Alpha vs R_Squared values
plt.scatter(x=alphas,y=R_sqrd);plt.xlabel("alpha");plt.ylabel("R_Squared")

# Alpha vs RMSE
plt.scatter(x=alphas,y=rmse);plt.xlabel("alpha");plt.ylabel("rmse")

# We got minimum R_Squared value at small alpha values 

# Running Lasso Regression
from sklearn.linear_model import Lasso
LM1 = Lasso(alpha = 0.01,normalize=True)
LM1.fit(x , y)

# Coefficient values for all the independent variables
LM1.coef_
LM1.intercept_

# Getting Adjusted R squared value
LM1.score(x,y)

pred1 = LM1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LM1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running a Ridge Regressor of set of alpha values and observing how the R-Squared
rmse = []
R_sqrd = []
alphas = np.arange(0,30,0.05)
for i in alphas:
    LM = Lasso(alpha = i,normalize=True)
    LM.fit(x , y)
    R_sqrd.append(LM.score(x , y))
    rmse.append(np.sqrt(np.mean((LM.predict(x) - y)**2)))
    
    
# Plotting RMSE,R_Squared values with respect to alpha values

# Alpha vs R_Squared values
plt.scatter(x=alphas,y=R_sqrd);plt.xlabel("alpha");plt.ylabel("R_Squared")

# Alpha vs RMSE
plt.scatter(x=alphas,y=rmse);plt.xlabel("alpha");plt.ylabel("rmse")

# We got minimum R_Squared value at small alpha values 

# So Simple Regression Model yields better results than Lasso and Ridge

import statsmodels.formula.api as smf
formula = "price" + "~" + "+".join(x.columns)
model = smf.ols(formula ,data=new_comp).fit() # regression model

# Summary
model.summary()


# influence index plots
import statsmodels.api as sm
fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model, alpha = 0.05 , ax = ax , criterion = "cooks") # Points 1440 , 1700 seems to be outliers

new_comp = new_comp.drop(new_comp.index[[1440,1700]] , axis=0)
model2 = smf.ols(formula ,data=new_comp).fit() # regression model

# Summary
model2.summary()
# Influce index plots
fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model2, alpha = 0.05 , ax = ax , criterion = "cooks") # Points 19, 24 ,27 seems to be outliers

#RMSE
pred2 = model2.predict(new_comp)
res = pred2 - new_comp.price
rmse = np.sqrt(np.mean(res * res))
rmse

new_comp = new_comp.drop(new_comp.index[[19 , 24, 27]] , axis=0)
model3 = smf.ols(formula ,data=new_comp).fit() # regression model
model3.summary()

#RMSE
pred3 = model3.predict(new_comp)
res = pred3 - new_comp.price
rmse = np.sqrt(np.mean(res * res))
rmse

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred3 , y = new_comp.price - pred3);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred3 - new_comp.price) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred3 , y = new_comp.price);plt.xlabel("Predicted");plt.ylabel("Actual")


# QQ plot for residuals 
import pylab          
import scipy.stats as st

# Checking Residuals are normally distributed
st.probplot(model3.resid_pearson, dist="norm", plot=pylab)


############ Homoscedasticity #######

# Residuals VS Fitted Values 
plt.scatter(pred3,model3.resid_pearson,c="r"),plt.axhline(y=0,color='blue');plt.xlabel("fitted_values");plt.ylabel("residuals")

# Spliting the data into train and test 
from sklearn.model_selection import train_test_split
train,test = train_test_split(new_comp,test_size=0.2)


# preparing the model on train data 

model_train = smf.ols(formula,data=train).fit()

# train_data prediction
train_pred = model_train.predict(train)

# train residual values 
train_resid  = train_pred - train.price

# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid*train_resid))

# prediction on test data set 
test_pred = model_train.predict(test)

# test residual values 
test_resid  = test_pred - test.price

# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid*test_resid))

# Both test and train RMSE are close by

File: Lasso - Ridge Regression/Computer_Data_Lasso_Ridge.R
library(readr)
comp <- read_csv(file.choose())

View(comp)

#Removing Index column as nothing can be predicted with this information
comp = comp[-1]

# Data Preparation & EDA

# checking for missing values

sum(is.na(comp)) # No missing values

# 6 - point Summary

summary(comp)
colnames(comp)

attach(comp)

library(dplyr)
summarise_if(comp , is.numeric, sd) # Standard Deviation
summarise_if(comp , is.numeric, max) - summarise_if(comp , is.numeric, min) # Range

# Skewness
#install.packages("moments")
library(moments)
skewness(comp[-c(6,7,8)]) # Excluding categorical features

#Kurtosis
kurtosis(comp[-c(6,7,8)])  # Excluding categorical features

#Graphical Representation

#Distribution
dens <- apply(comp[-c(6,7,8)] , 2 , density)
plot(NA, xlim=range(sapply(dens, "[", "x")), ylim=range(sapply(dens, "[", "y")))
mapply(lines, dens, col=1:length(dens))
legend("topright", legend=names(dens), fill=1:length(dens))

#Box Plot
boxplot(comp[-c(6,7,8)])

# Pair plot
pairs(comp[-c(6,7,8)])

# Correlation Matrix
panel.cor <- function(x, y, digits=2, prefix="", cex.cor)
{
  #usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r = (cor(x, y))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex <- 0.4/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex)
}
pairs(comp[-c(6,7,8)], upper.panel=panel.cor,main="Scatter Plot Matrix with Correlation Coefficients")


#Linear Model
model1 <- lm(price ~ . , data = comp)
summary(model1)

# Diagnostic Plots
#install.packages("car")
library(car)
plot(model1)# Residual Plots, QQ-Plos, Std. Residuals vs Fitted, Cook's distance


# Deletion Diagnostics for identifying influential variable
influence.measures(model1)
influenceIndexPlot(model1, id.n=3) 
influencePlot(model1, id.n=3) # Points 49 & 50 seems to be outliers



# Converting the data into compatible format in which model accepts 
x <- model.matrix(price~.-1,data=comp)
y <- comp$price

library(glmnet)

# setting lamda as 10^10 till 10^-2
lambda <- 10^seq(10, -2, length = 50)

# Ridge Regression
ridge_reg <- glmnet(x,y,alpha=0,lambda=lambda)
summary(ridge_reg)

# Coefficients vs Lambda
plot(ridge_reg,xvar="lambda",label=T)

# ridge regression coefficients, stored in a matrix 
dim(coef(ridge_reg))
plot(ridge_reg)

# Cross Validation in Ridge Regression
cv_ridge = cv.glmnet(x, y, alpha = 0) 

# Select lamda that minimizes training MSE
bestlam = cv_ridge$lambda.min  
bestlam

# Draw plot of training MSE as a function of lambda
plot(cv_ridge) 

ridge_pred1 = predict(ridge_reg, s = bestlam, newx = x)
sqrt(mean((ridge_pred1 - y)^2)) # RMSE is 278.6

ridge_pred1 = predict(ridge_reg, s = 0, newx = x)
sqrt(mean((ridge_pred1 - y)^2)) #RMSE is 275.12 indicates basic regression is better than lasso regression


# LASSO Regression
lasso_mod = glmnet(x,y, alpha = 1, lambda = lambda)

plot(lasso_mod)    # Draw plot of coefficients

cv_lasso = cv.glmnet(x, y, alpha = 1) # Fit lasso model on training data

plot(cv_lasso) # Draw plot of training MSE as a function of lambda

bestlam_lasso = cv_lasso$lambda.min # Select lamda that minimizes training MSE
bestlam_lasso
# Use best lambda to predict test data
lasso_pred = predict(lasso_mod, s = bestlam_lasso, newx = x)

sqrt(mean((lasso_pred - y)^2)) # RMSE is 275.15 which is is higher than basic multiple linear regression

# Display coefficients using lambda chosen by CV
lasso_coef = predict(lasso_mod, type = "coefficients", s = bestlam)[1:5,] 
lasso_coef


#######
# Partitioning Data into training set and testing set
# Spliting data into train and test
n=nrow(comp)
n1=n*0.7
n2=n-n1
train=sample(1:n,n1)
test = comp[-train,]

pred=predict(model1,newdat=test)
actual=test$price
error=actual-pred

test.rmse=sqrt(mean(error**2))
test.rmse

train.rmse = sqrt(mean(model1$residuals**2))
train.rmse 

# Train error and Test error are similar So the model is having a good fit